pybamm/expression_tree/jacobian.py

```diff
@@ -0,0 +1,46 @@
+#
+# Calculate the Jacobian of a symbol
+#
+
+
+class Jacobian(object):
+    def __init__(self, known_jacs=None):
+        self._known_jacs = known_jacs or {}
+
+    def jac(self, symbol, variable):
+        """
+        This function recurses down the tree, computing the Jacobian using
+        the Jacobians defined in classes derived from pybamm.Symbol. E.g. the
+        Jacobian of a 'pybamm.Multiplication' is computed via the product rule.
+        If the Jacobian of a symbol has already been calculated, the stored value
+        is returned.
+        Note: The Jacobian is the derivative of a symbol with respect to a (slice of)
+        a State Vector.
+
+        Parameters
+        ----------
+        symbol : :class:`pybamm.Symbol`
+            The symbol to calculate the Jacobian of
+        variable : :class:`pybamm.Symbol`
+            The variable with respect to which to differentiate
+
+        Returns
+        -------
+        :class:`pybamm.Symbol`
+            Symbol representing the Jacobian
+        """
+
+        try:
+            return self._known_jacs[symbol.id]
+        except KeyError:
+            jac = self._jac(symbol, variable)
+            self._known_jacs[symbol.id] = jac
+            return jac
+
+    def _jac(self, symbol, variable):
+        """ See :meth:`Jacobian.jac()`. """
+        jac = symbol._jac(variable)
+        # jacobian removes the domain(s)
+        jac.domain = []
+        jac.auxiliary_domains = {}
+        return jac
```

Comment:
In order to reuse already-calculated jacobians, this function needs to keep calling its own `jac` method. For example, for a binary operator, call `Jacobian.jac` on left and right, and then call `symbol._jac` on processed_left and processed_right:
```python3
processed_left = self.jac(symbol.left)
processed_right = self.jac(symbol.right)
return symbol._jac(processed_left, processed_right)
```
Otherwise a new `Jacobian` object just gets created each time. See the `Simplify` class for example.
Apologies if you're doing this in a different way that I've missed.

Reply:
ah, thanks -- I had misunderstood how `known_jacs` was getting passed around.